src/zcash/Address.cpp

```diff
@@ -71,3 +145,107 @@ uint32_t TypecodeForReceiver::operator()(
+    // Copy the C-string into C++.
+    std::string res(encoded);
+    // Free the C-string.
+    zcash_address_string_free(encoded);
+    return res;
+}
+
+std::optional<libzcash::SaplingDiversifiableFullViewingKey> libzcash::UnifiedFullViewingKey::GetSaplingKey() const {
+    std::vector<uint8_t> buffer(128);
+    if (unified_full_viewing_key_read_sapling(inner.get(), buffer.data())) {
+        CDataStream ss(buffer, SER_NETWORK, PROTOCOL_VERSION);
+        return SaplingDiversifiableFullViewingKey::Read(ss);
+    } else {
+        return std::nullopt;
+    }
+}
+
+std::optional<CChainablePubKey> libzcash::UnifiedFullViewingKey::GetTransparentKey() const {
+    std::vector<uint8_t> buffer(65);
+    if (unified_full_viewing_key_read_transparent(inner.get(), buffer.data())) {
+        CDataStream ss(buffer, SER_NETWORK, PROTOCOL_VERSION);
+        return CChainablePubKey::Read(ss);
+    } else {
+        return std::nullopt;
+    }
+}
+
+bool libzcash::UnifiedFullViewingKeyBuilder::AddTransparentKey(const transparent::AccountPubKey& key) {
+    if (t_bytes.has_value()) return false;
+    CDataStream ss(SER_NETWORK, PROTOCOL_VERSION);
+    ss << key.GetPubKey();
+    assert(ss.size() == 65);
+    std::vector<uint8_t> ss_bytes(ss.begin(), ss.end());
+    t_bytes = ss_bytes;
+    return true;
+}
+
+bool libzcash::UnifiedFullViewingKeyBuilder::AddSaplingKey(const SaplingDiversifiableFullViewingKey& key) {
+    if (sapling_bytes.has_value()) return false;
+    CDataStream ss(SER_NETWORK, PROTOCOL_VERSION);
+    ss << key;
+    assert(ss.size() == 128);
+    std::vector<uint8_t> ss_bytes(ss.begin(), ss.end());
+    sapling_bytes = ss_bytes;
+    return true;
+}
+
+std::optional<libzcash::UnifiedFullViewingKey> libzcash::UnifiedFullViewingKeyBuilder::build() const {
+    UnifiedFullViewingKeyPtr* ptr = unified_full_viewing_key_from_components(
+            t_bytes.has_value() ? t_bytes.value().data() : nullptr,
+            sapling_bytes.has_value() ? sapling_bytes.value().data() : nullptr);
+
+    if (ptr == nullptr) {
+        return std::nullopt;
+    } else {
+        return UnifiedFullViewingKey(ptr);
+    }
+}
+
+libzcash::UnifiedFullViewingKey libzcash::UnifiedFullViewingKey::FromZcashdUFVK(const ZcashdUnifiedFullViewingKey& key) {
+    UnifiedFullViewingKeyBuilder builder;
+    if (key.GetTransparentKey().has_value()) {
+        builder.AddTransparentKey(key.GetTransparentKey().value());
+    }
+    if (key.GetSaplingKey().has_value()) {
+        builder.AddSaplingKey(key.GetSaplingKey().value());
+    }
+
+    auto result = builder.build();
+    if (!result.has_value()) {
+        throw std::invalid_argument("Cannot convert from invalid viewing key.");
+    }
+    return result.value();
+}
+
+libzcash::UFVKId libzcash::UnifiedFullViewingKey::GetKeyID(const KeyConstants& keyConstants) const {
+    // The ID of a ufvk is the blake2b hash of the serialized form of the
+    // ufvk with the receivers sorted in order of descending receiver type.
```

Comment:
```suggestion
    // ufvk with the receivers sorted in typecode order.
```

Reply:
Addressed in https://github.com/zcash/zcash/pull/5543